localstack/services/apigateway/integration.py

```diff
@@ -0,0 +1,49 @@
+import json
+import logging
+
+from localstack import config
+from localstack.services.apigateway.apigateway_listener import apply_template
+from localstack.services.apigateway.context import ApiInvocationContext
+from localstack.utils.aws import aws_stack
+from localstack.utils.common import make_http_request, to_str
+
+LOG = logging.getLogger(__name__)
+
+
+class BackendIntegration:
+    """
+    Backend integration
+    """
+
```

Comment:
Should we add `invoke(...)` as an abstract method to the base class, to make it a bit more clear/explicit what the interface of `BackendIntegration`s looks like?

Also, wondering if we can make the `invocation_context` part of the signature of the `invoke(..)` method (and remove it from the `__init__` initializer), as the `invocation_context` is indeed tightly coupled to the invocation itself. In other words, the concrete `BackendIntegration` instance feels more like a static property of the REST API itself (as defined by the `integrationUri`), whereas the `invocation_context` is an inherent part of the invocation via `invoke(..)`. Also, this would give us a bit more flexibility over time, as we could be using the same `BackendIntegration` instance for multiple consecutive invocations.

Reply:
So I left it open in purpose, I have a branch with the backend integrations refactoring and haven't settled in the API.

My thoughts, as highlighted above, were to include on each integration the request templating and response templating - this is transversal to all integrations. As having the `invocation_context` passed to `invoke` seems a good option.

Now, the `integrationUri`, and that might explain my option (as I said not settled), was to create an instance of `BackendIntegration` based on the `uri`, more like an abstract factory.

Nevertheless, I'm not doing in this PR  - let's focus these conversations on this one https://github.com/localstack/localstack-ext/compare/master...backend_integration_refactor

Reply:
> Nevertheless, I'm not doing in this PR - let's focus these conversations on ...

Excellent, let's do that 👍